t/js_saml.t

```diff
@@ -0,0 +1,1642 @@
+    }
+
+    return $r;
+}
+
+sub cfg_post {
+    my ($arg, $post, $host) = @_;
+    $host //= 'sp.example.com';
+    my $json;
+
+    if (ref $arg eq 'HASH') {
+        $json = $arg;
+    } elsif (!ref $arg) {
+        my ($key, $value) = each %{decode_json($arg)};
+        $cfg->{$key} = $value;
+        $json = {$key => $value};
+    } else {
+        die "Invalid arguments for cfg_post";
+    }
+
+    for my $key (keys %$json) {
+        my $value = $json->{$key};
+        my $data = { $host => $value };
+        my $data_string = encode_json($data);
+
+        if ($post) {
+            http_post("$kv/$key", $data_string);
+        } else {
+            http_patch("$kv/$key", $data_string);
+        }
+    }
+}
+
+sub cfg_verify {
+    my ($param, $test_name) = @_;
+
+    my $url = "$kv/$param";
+    my $original_value = getkv($url);
+
+    http_patch($url, '{"sp.example.com": ""}');
+    my $r = get('/login');
+    http_patch($url, $original_value);
+
+    like($r, qr/(?=.*Invalid)(?=.*$param)/s, $test_name);
+
+    return $r;
+}
+
+sub http_post {
+	my ($url, $body, %extra) = @_;
+	my $len = length($body);
+
+    my $auth_token = $extra{auth_token} || '';
+    my $auth_redir = $extra{auth_redir} || '';
+
+	http(<<EOF);
+POST $url HTTP/1.0
+Host: sp.example.com
+Content-Length: $len
+Cookie: auth_token=$auth_token
+Cookie: auth_redir=$auth_redir
+Content-Type: application/x-www-form-urlencoded
+
+$body
+EOF
+}
+
+sub http_patch {
+	my ($url, $body) = @_;
+	my $len = length($body);
+
+	http(<<EOF);
+PATCH $url HTTP/1.1
+Host: sp.example.com
+Connection: close
+Content-Length: $len
+
+$body
+EOF
+}
+
+sub get {
+    my ($uri, %extra) = @_;
+
+    my $auth_token = $extra{auth_token} || '';
+
+    http(<<EOF);
+GET $uri HTTP/1.0
+Host: sp.example.com
+Cookie: auth_token=$auth_token
+
+EOF
+}
+
+sub recode {
+	my $json = JSON::PP::decode_json(shift);
+	JSON::PP->new()->canonical()->encode($json);
+}
+
+sub getkv {
+	my ($uri) = @_;
+
+	get($uri) =~ /\x0d\x0a?\x0d\x0a?(.*)/ms;
+	recode($1);
+}
+
+sub api {
+	my ($uri, %extra) = @_;
+
+	$uri = defined $uri ? "/api/$api_version$uri" : '/api';
+	my ($body) = http_get($uri, %extra) =~ /.*?\x0d\x0a?\x0d\x0a?(.*)/ms;
+
+	return decode_json($body);
+}
+
+sub validate_saml_signature {
+    my ($xmlDoc, $public_key_pem) = @_;
+
+    my $xpc = XML::LibXML::XPathContext->new($xmlDoc);
+    $xpc->registerNs('saml', 'urn:oasis:names:tc:SAML:2.0:assertion');
+    $xpc->registerNs('samlp', 'urn:oasis:names:tc:SAML:2.0:protocol');
+    $xpc->registerNs('ds', 'http://www.w3.org/2000/09/xmldsig#');
+
+    my ($signature_node) = $xpc->findnodes('//ds:Signature');
+    return 0 unless $signature_node;
+
+    my ($signed_info_node) = $xpc->findnodes('./ds:SignedInfo',
+        $signature_node);
+    my ($signature_method_node) = $xpc->findnodes('./ds:SignatureMethod',
+        $signed_info_node);
+    my $signature_algorithm = $signature_method_node->
+        getAttribute('Algorithm');
```

Comment:
You don't have to split this into two lines.

Reply:
I do, it's 81 chars